File: android/lib/src/main/kotlin/blake/hash/Blake3.kt
package blake.hash

import blake.hash.internal.BLAKE3Core
import blake.hash.internal.BLAKE3Core.CHUNK_LEN
import blake.hash.internal.BLAKE3Core.DERIVE_KEY_CONTEXT
import blake.hash.internal.BLAKE3Core.DERIVE_KEY_MATERIAL
import blake.hash.internal.BLAKE3Core.IV
import blake.hash.internal.BLAKE3Core.KEYED_HASH
import blake.hash.internal.BLAKE3Core.KEY_LEN
import blake.hash.internal.BLAKE3Core.OUT_LEN

/**
 * BLAKE3 cryptographic hash function.
 *
 * Supports hash, keyed hash (MAC/PRF), and key derivation modes,
 * with extendable output (XOF).
 */
public class BLAKE3 private constructor() {

    public companion object {
        /** Hash mode — returns 32-byte digest. */
        public fun hash(input: ByteArray): ByteArray {
            return Hasher().update(input).finalize()
        }

        /** Keyed hash mode (MAC/PRF) — key must be exactly 32 bytes. */
        public fun keyedHash(key: ByteArray, input: ByteArray): ByteArray {
            return Hasher(key).update(input).finalize()
        }

        /** Key derivation mode — returns 32-byte derived key. */
        public fun deriveKey(context: String, keyMaterial: ByteArray): ByteArray {
            return Hasher.deriveKey(context).update(keyMaterial).finalize()
        }
    }

    /**
     * Incremental BLAKE3 hasher supporting streaming updates and XOF output.
     */
    public class Hasher private constructor(
        private val key: IntArray,
        private val baseFlags: Int
    ) {
        private var chunkState = BLAKE3Core.ChunkState(key, 0, baseFlags)
        private val cvStack = Array(54) { IntArray(8) } // max tree depth
        private var cvStackLen = 0
        private var chunkCount: Long = 0

        /** Hash mode. */
        public constructor() : this(IV.copyOf(), 0)

        /** Keyed hash mode — key must be exactly 32 bytes. */
        public constructor(key: ByteArray) : this(keyWords(key), KEYED_HASH)

        public companion object {
            /** Create a derive_key hasher (phase 2). */
            public fun deriveKey(context: String): Hasher {
                val contextBytes = context.encodeToByteArray()
                val contextHasher = Hasher(IV.copyOf(), DERIVE_KEY_CONTEXT)
                contextHasher.update(contextBytes)
                val contextKey = contextHasher.rootOutput().rootChainingValue()
                return Hasher(contextKey, DERIVE_KEY_MATERIAL)
            }

            private fun keyWords(key: ByteArray): IntArray {
                require(key.size == KEY_LEN) { "Key must be exactly $KEY_LEN bytes" }
                return IntArray(8) { BLAKE3Core.leToInt(key, it * 4) }
            }
        }

        public fun update(input: ByteArray): Hasher = update(input, 0, input.size)

        public fun update(input: ByteArray, offset: Int, length: Int): Hasher {
            var pos = offset
            var remaining = length
            while (remaining > 0) {
                if (chunkState.isComplete) {
                    completeChunk()
                }
                val take = minOf(CHUNK_LEN - chunkState.bytesConsumed, remaining)
                chunkState.update(input, pos, take)
                pos += take
                remaining -= take
            }
            return this
        }

        /** Finalize with default 32-byte output. */
        public fun finalize(): ByteArray = finalizeXof(OUT_LEN)

        /** Finalize with arbitrary output length (XOF). */
        public fun finalizeXof(outputLength: Int): ByteArray {
            return rootOutput().rootOutputBytes(outputLength)
        }

        // ---- Internal tree management ----

        private fun completeChunk() {
            val chunkCv = chunkState.output().chainingValueWords()
            chunkCount++
            addChunkCv(chunkCv)
            chunkState = BLAKE3Core.ChunkState(key, chunkCount, baseFlags)
        }

        /**
         * After pushing a chunk's chaining value, merge adjacent pairs that
         * form complete subtrees. A complete subtree at level L exists when
         * bit L of the total chunk count is 0 (the bit just cleared).
         */
        private fun addChunkCv(cv: IntArray) {
            cv.copyInto(cvStack[cvStackLen])
            cvStackLen++
            var totalChunks = chunkCount
            while (cvStackLen > 1 && (totalChunks and 1L) == 0L) {
                cvStackLen--
                val right = cvStack[cvStackLen]
                cvStackLen--
                val left = cvStack[cvStackLen]
                val parent = BLAKE3Core.parentChainingValue(left, right, key, baseFlags)
                parent.copyInto(cvStack[cvStackLen])
                cvStackLen++
                totalChunks = totalChunks ushr 1
            }
        }

        private fun rootOutput(): BLAKE3Core.Output {
            // Finalize the current (possibly partial) chunk
            var output = chunkState.output()
            var cv = output.chainingValueWords()

            // If there are stacked CVs, merge right-to-left
            var idx = cvStackLen - 1
            if (idx < 0) {
                // Single chunk — it is the root
                return output
            }

            // The current chunk is the rightmost child
            while (idx > 0) {
                cv = BLAKE3Core.parentChainingValue(cvStack[idx], cv, key, baseFlags)
                idx--
            }

            // Last merge gets ROOT flag via parentOutput
            return BLAKE3Core.parentOutput(cvStack[0], cv, key, baseFlags)
        }
    }
}

File: ios/Sources/BlakeHash/Blake3.swift
import Foundation

/// BLAKE3 cryptographic hash function — public API.
/// Pure Swift implementation with zero external dependencies.
public struct BLAKE3: Sendable {

    // MARK: - Convenience functions

    /// Hash the input with BLAKE3 (32-byte output).
    public static func hash(_ input: Data) -> Data {
        var hasher = Hasher()
        hasher.update(input)
        return hasher.finalize()
    }

    /// Keyed hash (MAC/PRF). Key must be exactly 32 bytes.
    public static func keyedHash(key: Data, data: Data) -> Data {
        var hasher = Hasher(key: key)
        hasher.update(data)
        return hasher.finalize()
    }

    /// Derive a 32-byte key from context string and key material.
    public static func deriveKey(context: String, keyMaterial: Data) -> Data {
        var hasher = Hasher.deriveKey(context: context)
        hasher.update(keyMaterial)
        return hasher.finalize()
    }

    // MARK: - Incremental Hasher

    public struct Hasher: Sendable {
        private var engine: BLAKE3Engine

        /// Hash mode (unkeyed).
        public init() {
            engine = BLAKE3Engine(key: BLAKE3Constants.IV, flags: 0)
        }

        /// Keyed hash mode. Key must be exactly 32 bytes.
        public init(key: Data) {
            let keyBytes = [UInt8](key)
            precondition(keyBytes.count == 32, "BLAKE3 keyed hash requires a 32-byte key")
            let keyWords = wordsFromBytes(keyBytes)
            engine = BLAKE3Engine(key: keyWords, flags: BLAKE3Constants.KEYED_HASH)
        }

        /// Internal init for derive_key phase 2.
        private init(derivedKey: [UInt32]) {
            engine = BLAKE3Engine(key: derivedKey, flags: BLAKE3Constants.DERIVE_KEY_MATERIAL)
        }

        /// Create a derive_key hasher.
        public static func deriveKey(context: String) -> Hasher {
            // Phase 1: hash the context string with DERIVE_KEY_CONTEXT
            let contextBytes = Array(context.utf8)
            var contextHasher = BLAKE3Engine(
                key: BLAKE3Constants.IV,
                flags: BLAKE3Constants.DERIVE_KEY_CONTEXT
            )
            contextHasher.update(contextBytes)
            let contextOutput = contextHasher.finalOutput()
            // Use first 8 words (32 bytes) of root output as the derived key
            let contextKeyBytes = contextOutput.rootOutputBytes(outputLength: BLAKE3Constants.KEY_LEN)
            let contextKey = wordsFromBytes(contextKeyBytes)

            // Phase 2: the returned hasher uses DERIVE_KEY_MATERIAL
            return Hasher(derivedKey: contextKey)
        }

        /// Feed input data incrementally.
        public mutating func update(_ input: Data) {
            engine.update([UInt8](input))
        }

        /// Finalize with default 32-byte output.
        public func finalize() -> Data {
            Data(engine.finalOutput().rootOutputBytes(outputLength: BLAKE3Constants.OUT_LEN))
        }

        /// Finalize with arbitrary output length (XOF).
        public func finalizeXof(outputLength: Int) -> Data {
            Data(engine.finalOutput().rootOutputBytes(outputLength: outputLength))
        }
    }
}
